### ISSUE TITLE
To copy/paste some code from the diff view doesn't include empty lines anymore

### ISSUE THREAD

**Issue by agarny** [desktop/desktop#17652]
### The problem

I have a diff that contains some empty lines. When I used to copy/paste such a diff, I would (rightly) get the empty lines. However, since upgrading to version 3.3.5, I don't get those empty lines anymore.

### Release version

3.3.5

### Operating system

macOS 14.1

### Steps to reproduce the behavior

1. Make some changes to a file including deleting some empty lines;
2. In the diff view of GitHub Desktop, copy the code that you have just deleted in step 1; and
3. Paste the code you have copied in step 2.

After step 3, you will see the copied code, but not the empty lines. For instance, say that you deleted the following code:
```
// First line.
// Second line, followed by an empty line.

// Fourth line, preceded by an empty line.
```
Top copy/paste the diff will result in the following code:
```
// First line.
// Second line, followed by an empty line.
// Fourth line, preceded by an empty line.
```
i.e. the empty line is gone while it used to be (rightly) included.

**Comment by agarny** [comment]
FWIW, to copy/paste one line from the diff view doesn't include the new line character either.

**Comment by niik** [comment]
Hey @agarny, thanks for reporting this!

I've been able to reproduce it and I'm assuming it's related to the new diff engine (https://github.com/desktop/desktop/commit/790fb88bf9042031bcf54661effd427b444fdec7). Going to take a closer look.

**Comment by agarny** [comment]
Thanks @niik. Looking at your PR https://github.com/desktop/desktop/pull/17653, it seems like it's only for empty lines? As mentioned in my follow-up comment, the issue also arises when copying/pasting one line. Say that you have the following code:
```
Line 1
Line 2
Line 3
```
and the following diff:
```
Some diff
```
Now, say that you copy/paste that one liner between `Line 1` and `Line 2` then you will get the following:
```
Line 1
Some diffLine 2
Line 3
```
while I would expect (as before) to get:
```
Line 1
Some diff
Line 2
Line 3
```

**Comment by niik** [comment]
Hey @agarny, sorry, I missed your comment about including a trailing newline. Having considered this for a bit I don't think that's the behavior most would expect. The diff rendering backend we used previously (CodeMirror) used to do this which make more sense there since it's a text editor but I don't see a compelling reason to diverge from the normal text copying semantics in Chromium here.

As @tidy-dev points out in #17653 our current behavior also aligns with the diff viewer behavior on GitHub.com (i.e we don't include the newline when copying a single line).

**Comment by tidy-dev** [comment]
The fix for this has shipped to beta. You can download it at https://desktop.github.com/beta/ if you like the fix before it moves into the next production release.

**Comment by agarny** [comment]
> Hey @agarny, sorry, I missed your comment about including a trailing newline. Having considered this for a bit I don't think that's the behavior most would expect. The diff rendering backend we used previously (CodeMirror) used to do this which make more sense there since it's a text editor but I don't see a compelling reason to diverge from the normal text copying semantics in Chromium here.

Don't see a compelling reason? Why would I want to copy some text from the diff view if it's not to paste it some kind of text editor, which means that I would expect the full line to be copied and that includes the trailing newline, not to mention that this is how GitHub Desktop used to work before. Disappointing.

**Comment by agarny** [comment]
> The fix for this has shipped to beta. You can download it at https://desktop.github.com/beta/ if you like the fix before it moves into the next production release.

Thanks, it's working as per the PR. I don't like it though and I have just reverted to version 3.3.4 where copying/pasting works as I would really expect. (Yeah, I don't like regressions.)

**Comment by agarny** [comment]
Actually, @niik, if you check your [PR](https://github.com/desktop/desktop/pull/17653/files), it's the same story. I mean, if I have:
```
Line 1
Line 2
Line 3
```
and copy line 362 of your PR and paste it at the beginning of `Line 3`, I end up with:
```
Line 1
Line 2
          {data.content.length === 0 && <br />}
Line 3
```
and not:
```
Line 1
Line 2
          {data.content.length === 0 && <br />}Line 3
```
as is the case with your current fix. So, yeah, not consistent between GitHub Desktop and the web interface.

**Comment by niik** [comment]
> So, yeah, not consistent between GitHub Desktop and the web interface.

Ah, I was initially unable to reproduce this but I think I found an explanation. The split-diff viewer on GitHub.com (which I use) does not include the line break whereas the unified diff does. So there's no real consistency to speak about here. I've looked at the file viewer on GitHub.com (as opposed to the diff viewer) and it will copy newlines. I've also looked at StackOverflow and MDN as examples of places where one is likely to copy code from and they also share the behavior of including the newline when triple-clicking a line.

I think the above coupled with the divergence from earlier versions of Desktop tip the scale in favor of attempting to reinstate the previous behavior.

**Comment by agarny** [comment]
> I think the above coupled with the divergence from earlier versions of Desktop tip the scale in favor of attempting to reinstate the previous behavior.

Thanks a lot, that would be very nice. 🙏 (Not least since I really don't like to see that `An updated version of GitHub Desktop...` message. :)

**Comment by agarny** [comment]
Any chance of getting access to a beta version with this fix? From what I can tell, the current beta version is 3.3.6-beta2 from 2 November and it (obviously) doesn't include this fix.

**Comment by tidy-dev** [comment]
@agarny This is shipped to in the latest beta version `3.3.6-beta3`. Thank you for your patience.

**Comment by agarny** [comment]
Thanks @tidy-dev. Yes, I noticed the new beta and I have been using it successfully. Thanks @niik for fixing that issue, much appreciated!
